servicetalk-loadbalancer/src/main/java/io/servicetalk/loadbalancer/LoadBalancerBuilder.java

```diff
@@ -0,0 +1,123 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.servicetalk.loadbalancer;
+
+import io.servicetalk.client.api.LoadBalancedConnection;
+import io.servicetalk.client.api.LoadBalancer;
+import io.servicetalk.client.api.LoadBalancerFactory;
+import io.servicetalk.client.api.ServiceDiscoverer;
+import io.servicetalk.concurrent.api.Executor;
+
+import java.time.Duration;
+
+interface LoadBalancerBuilder<ResolvedAddress, C extends LoadBalancedConnection> {
+    /**
+     * Set the {@code loadBalancingPolicy} to use with this load balancer.
+     * @param loadBalancingPolicy the policy to use
+     * @return {@code this}
+     */
+    LoadBalancerBuilder<ResolvedAddress, C> loadBalancingPolicy(LoadBalancingPolicy loadBalancingPolicy);
+
+    /**
+     * This {@link LoadBalancer} may monitor hosts to which connection establishment has failed
+     * using health checks that run in the background. The health check tries to establish a new connection
+     * and if it succeeds, the host is returned to the load balancing pool. As long as the connection
+     * establishment fails, the host is not considered for opening new connections for processed requests.
+     * If an {@link Executor} is not provided using this method, a default shared instance is used
+     * for all {@link LoadBalancer LoadBalancers} created by this factory.
+     * <p>
+     * {@link #healthCheckFailedConnectionsThreshold(int)} can be used to disable this mechanism and always
+     * consider all hosts for establishing new connections.
+     *
+     * @param backgroundExecutor {@link Executor} on which to schedule health checking.
+     * @return {@code this}.
+     * @see #healthCheckFailedConnectionsThreshold(int)
+     */
+    LoadBalancerBuilder<ResolvedAddress, C> backgroundExecutor(Executor backgroundExecutor);
+
+    /**
+     * Sets the linear search space to find an available connection for the next host.
+     * <p>
+     * When the next host has already opened connections, this {@link LoadBalancer} will perform a linear search for
+     * a connection that can serve the next request up to a specified number of attempts. If there are more open
+     * connections, selection of remaining connections will be attempted randomly.
+     * <p>
+     * Higher linear search space may help to better identify excess connections in highly concurrent environments,
+     * but may result in slightly increased selection time.
+     *
+     * @param linearSearchSpace the number of attempts for a linear search space, {@code 0} enforces random
+     * selection all the time.
+     * @return {@code this}.
+     */
+    LoadBalancerBuilder<ResolvedAddress, C> linearSearchSpace(int linearSearchSpace);
```

Comment:
Should this be moved in the `RoundRobin` policy builder?

Reply:
No, it's actually used by `DefaultHost` as it iterates over open connections to a particular host, not by the host selection level. At some point it might be good to abstract away the pooling part of host as well at some point but it's not a major priority.